server/raft.go

```diff
@@ -2922,7 +2915,7 @@ func (n *raft) runAsCandidate() {
 	// We vote for ourselves.
 	votes := 1
 
-	for {
+	for n.State() == Candidate {
```

Comment:
We tried this before with the for loops and I thought it presented issues no?

Reply:
We did have these in before, yes, and we reverted the non-leader loops in #4725 because we thought it had broken observer mode, but it was later proven to be an unrelated change at fault there. We just never put them back in the end.

Reply:
Make reading easier, I was struggling til I saw what the bool meant way down in the PR. Meaning when we go back to this code after some time away will have similar issue IMO.

Reply:
Do you recall what the unrelated change was?

Reply:
#4727 — we initially thought that the `Observer` state was broken by the condition on the `runAsFollower` loop, but it turned out that the `Observer` state const was never used. The unit test I added in that PR proved it still worked after we cleaned up the unused `Observer` state const.